### DISCUSSION TITLE
About the binary size of compiled product using Diesel

### DISCUSSION THREAD

**Discussion by SKTT1Ryze** [diesel-rs/diesel#2958] [Q&A] [answered]
> ref: https://github.com/SeaQL/sea-orm/issues/308

I'm writing Rust code in a company and we use diesel to build our project.  
The problem occurs on our project in diesel is that it brings too much binary size of compiled product.  
When I add one diesel related code like:  
```
fn a() {
    use schema::users::dsl::*;
    diesel::update(users.filter(id.eq(1)))
        .set(names.eq("ccc"))
        .get_result(&connection);
}
```
The increment of the compiled product size may be about 1kb. I don't know the exact number, but it does bring larger binary size than common Rust code. In production environment, we must concern about that problem.  

I guess that diesel write too much `generic code`(I don't mean diesel shouldn't), which may cause large binary size.  

May I ask some solutions about this?

**Comment by weiznich** [comment, marked as answer]
Thanks for opening this thread. I can understand your concerns that the use of generics can cause the size of compiled code to explode. That's generally true, at least if generics are used in a certain way. For diesel that's not that simple. I did a simple experiment here to demonstrate this. For this I wrote a simple small script to generate rust programs using diesel. It generates a file containing x times the function you provided above, where x can be passed via command line. Every function is called from main, so that we can be sure the compiler does not strip it out.

<details>
<summary> Script to generate rust programs using diesel </summary>

```python
#!/usr/bin/env python3

import sys

out = sys.argv[1]
count = int(sys.argv[2])

main = open(out, "w")

main.write('''#[macro_use]
extern crate diesel;

use diesel::prelude::*;

table! {
    users(id) {
        id -> Integer,
        names -> Text,
    }
}

''');

for i in range(0, count):
    main.write("fn update_" + str(i));
    main.write('''(connection: &PgConnection) {
    use self::users::dsl::*;

    diesel::update(users.filter(id.eq(1)))
        .set(names.eq("ccc"))
        .get_result::<(i32, String)>(connection);
}

''');

main.write('''
fn main() {
    let conn = PgConnection::establish("").unwrap();

''');

for i in range(0, count):
    main.write("    update_" + str(i) + "(&conn);\n");

main.write("}");
main.close();

```

</details>

Now given your report the hypothesis is that if you have that function 1000 times there we should see a binary size increase of ~ 1kb * 1000 = 1Mb, which should be clearly measurable.

Now I used the following configuration to build two example programs (one with 1000 function calls, one with a single function call):

* Rust: 1.56.1 
* Diesel: 1.4.8 (Postgres backend)
* Binaries are build with `cargo build --release`, without any special profile configuration, just cargos defaults

This results in the following binary sizes:
* For 1 function: 3747648 (3,6 MB)
* For 1000 functions: 3870824 (3,7MB)

That means we observe here a binary size increase of 0,1MB (or ~150 KB) instead of the predicted 1 MB, which is smaller by a factor of 10. 
Next I've looked at which part of the binary where responsible for the size increase. I've used [`cargo-bloat`](https://github.com/RazrFalcon/cargo-bloat/) for this. Again I've build and compared the release versions:

* For 1 function: 
```
cargo bloat --release --crates                                                                      
   Compiling diesel_size_test v0.1.0 (/tmp/diesel_size_test)
    Finished release [optimized] target(s) in 1.72s
    Analyzing target/release/diesel_size_test

File  .text     Size Crate
5.9%  90.5% 217.4KiB std
0.5%   7.1%  17.1KiB diesel
0.0%   0.2%     609B diesel_size_test
0.0%   0.0%      99B [Unknown]
6.6% 100.0% 240.3KiB .text section size, the file size is 3.6MiB
```
* For 1000 functions:
```
cargo bloat --release --crates
   Compiling diesel_test v0.1.0 (/tmp/diesel_size_test)
    Finished release [optimized] target(s) in 7.60s
    Analyzing target/release/diesel_size_test

File  .text     Size Crate
5.8%  60.3% 217.5KiB std
3.2%  33.5% 120.6KiB diesel_size_test
0.5%   4.7%  17.1KiB diesel
0.0%   0.0%      99B [Unknown]
9.5% 100.0% 360.5KiB .text section size, the file size is 3.7MiB
```

According to this numbers the code size generated by the test crate using diesel increased by ~120kb, while all other sizes remained exactly the same. Now this means using a lot of diesel queries does not result in a increased size of diesel itself. Now one could guess that that size increase is caused by code generated by diesel inside our test crate. So let's check what's a "normal" size increase for the same number of functions there by just using something other (Opening  a file in this case, as this should have similar complexity but no generic code) 

<details>
<summary> Script to generate the non-diesel test code </summary>

```python
#!/usr/bin/env python3

import sys

out = sys.argv[1]
count = int(sys.argv[2])

main = open(out, "w")

main.write('''
use std::fs::File;
use std::io::prelude::*;

''');
for i in range(0, count):
    main.write("#[inline(never)] fn update_" + str(i));
    main.write('''() {
    let mut f = File::open("/tmp/foo").unwrap();
}

''');

main.write('''
fn main() {

''');

for i in range(0, count):
    main.write("    update_" + str(i) + "();\n");

main.write("}");
main.close();

```
</details>

Results via `cargo bloat --release --crates`:

* For a single function: 
```
cargo bloat --release --crates 
   Compiling diesel_test v0.1.0 (/tmp/diesel_size_test)
    Finished release [optimized] target(s) in 0.43s
    Analyzing target/release/diesel_size_test

File  .text     Size Crate
5.7%  98.1% 204.7KiB std
0.0%   0.1%     154B diesel_size_test
0.0%   0.0%      99B [Unknown]
5.8% 100.0% 208.7KiB .text section size, the file size is 3.5MiB
```

* For 1000 functions:
```
cargo bloat --release --crates  
   Compiling diesel_test v0.1.0 (/tmp/diesel_size_test)
    Finished release [optimized] target(s) in 7.93s
    Analyzing target/release/diesel_size_test

File  .text     Size Crate
5.2%  55.4% 204.7KiB std
3.8%  40.7% 150.4KiB diesel_size_test
0.0%   0.0%      99B [Unknown]
9.3% 100.0% 369.7KiB .text section size, the file size is 3.9MiB
```

For this example the binary size is increased by ~150kb instead of ~120kb. At least I would interpret that as "Increasing the number of functions doing some work by the factor of 1000 will increase the binary size by 100 - 200kb" or something like that. I hope that this demonstrates that this is not caused by generic code bloat like you guessed.

Now let's try to explain why diesel does not increase the binary size via generic code bloat for each query:

Most of the types in diesel are so called zero sized structs, which either do not contain any fields at all, or where each field is a zero sized struct by itself. That means `std::mem::size_of` returns zero for those types. Especially most of diesels queries are represented by such types. Because they do not have any size at all they cannot be present in the final binary, they are only a compile time construct. Rustc/LLVM removes them while producing the final binary. As consequence of this rustc is forced to inline and optimize away basically all of diesels query builder. What's not there cannot be cause bloat.

Hopefully this explains the issue quite well. If there are any other question about this feel free to leave a message here, if you are able to produce code bloat via diesels generic methods I would be interested in a complete example showing this.

**Reply by SKTT1Ryze** [reply]
I appreciate your help very much.  
It' s very kind of you to do such an experiment to explain the issue, and I can learn much from that.  
I understand that diesel code does not increase the binary size because of most of types in diesel are zero sized, and I would like to call it well programming.  
Finally thanks for your patience again.  
> Please forgive my bad english. : )
